### PR TITLE
Clean up autocmds only when needed

### LINKED ISSUE #17561: `autocmd!` speed regression with api-autocmds
### Neovim version (nvim -v)

NVIM v0.7.0-dev+1155-g991e47288

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

Ubuntu 20.04

### Terminal name/version

GNOME Terminal 3.36.2 using VTE 0.60.3 +BIDI +GNUTLS +ICU +SYSTEMD

### $TERM environment variable

tmux-256color

### Installation

build from repo

### How to reproduce the issue

Startup time:
```
nvim --clean --startuptime log --cmd 'filetype plugin indent on'
```

Profile:
```
nvim --clean --cmd 'profile start profile.log' --cmd 'profile func *' --cmd 'profile file *' --cmd 'filetype plugin indent on' -c 'profdel func *' -c 'profdel file *' -c 'qa!'
```

### Expected behavior

Before #14661
* startup time 15.923 seconds
* `autocmd` 0.00697 seconds, `autocmd!` **0.000242** seconds

The raw data is here: https://gist.github.com/tomtomjhj/84285e43ab95496244bd343d44e3281e

### Actual behavior

#14661
* startup time 039.874 seconds
* `autocmd` 0.005301 seconds, `autocmd!` **0.024399** seconds

### PR DESCRIPTION
**Pull request by tomtomjhj**
Fixes #17561.

`au_need_clean` should be set only when there's a group to be cleared, and this is done by `aupat_del`.

### CODE REVIEW

**Review by lewis6991 — APPROVED** [review]

File: src/nvim/autocmd.c [hunk 1]
```diff
@@ -239,7 +239,6 @@ void aupat_del_for_event_and_group(event_T event, int group)
     }
   }
 
-  au_need_clean = true;
   au_cleanup();  // may really delete removed patterns/commands now
```
Comment (zeertzjq) [inline src/nvim/autocmd.c:242]: BTW, I think this comment look a bit strange.

Comment (zeertzjq) [inline src/nvim/autocmd.c:242]: This comment seems to be in the wrong place, but this is unrelated to this PR.

**Review by zeertzjq — APPROVED** [review]
